lib/srv/app/session.go

```diff
@@ -140,8 +169,54 @@ func (s *Server) newSessionChunk(ctx context.Context, identity *tlsca.Identity,
 	return sess, nil
 }
 
+// acquire() increments in-flight request count by 1.
+// It is supposed to be paired with a `release()` call,
+// after the chunk is done with for the individual request
+func (s *sessionChunk) acquire() error {
+	s.inflightCond.L.Lock()
+	defer s.inflightCond.L.Unlock()
+
+	if s.inflight == -1 {
+		return trace.Wrap(errSessionChunkAlreadyClosed)
+	}
+
+	s.inflight++
+	return nil
+}
+
+func (s *sessionChunk) release() {
+	s.inflightCond.L.Lock()
+	defer s.inflightCond.L.Unlock()
+	if s.inflight == -1 {
+		return
+	}
+	s.inflight--
+	s.inflightCond.Signal()
+}
+
 func (s *sessionChunk) close(ctx context.Context) error {
+	deadline := time.Now().Add(s.closeTimeout)
+	t := time.AfterFunc(s.closeTimeout, func() { s.inflightCond.Signal() })
+	defer t.Stop()
+
+	// Wait until there are no requests in-flight,
+	// then mark the session as not accepting new requests,
+	// and close it.
+	s.inflightCond.L.Lock()
```

Comment:
If possible, add a test for this case: inflight gets released during close wait.

Reply:
:+1.  first time seeing the `Cond` syntax. was a little weird to me.

Reply:
Condition variables usually behave this way: `Lock()`, check your conditions under the lock (since you want consistent state), then call `Wait()`, which unlocks so other threads can freely `Lock(); Notify()`. Before `Wait()` returns to the caller, it re-locks again. So in your code, you only "lock once" and "unlock once".

See also the docs for `sync` package:

> Wait atomically unlocks c.L and suspends execution of the calling goroutine. After later resuming execution, Wait locks c.L before returning.
> https://pkg.go.dev/sync#Cond.Wait

---

In theory, `TestSessionChunkCloseNormal` covers this case, as it spawns `close()` in a background goroutine, and only then decrements the counter.

https://github.com/gravitational/teleport/blob/ee9932984bfdde6f821d5cd5082b0d516d516449/lib/srv/app/session_test.go#L49-L50

However, it is not guaranteed that `close()` will proceed all the way to blocking on `Wait()` before we try to `release()`. I am not sure if there is a clean way to guarantee the ordering of these statements in the test. If you have an idea on how to do that nicely, please let me know :slightly_smiling_face: